servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingServiceDiscoverer.java

```diff
@@ -0,0 +1,149 @@
+            final ServiceDiscovererEventsCache<R, E> eventsCache =
+                    new ServiceDiscovererEventsCache<>(targetResource, makeUnavailable);
+            return delegate().discover(address)
+                    .map(eventsCache::consumeAndFilter)
+                    .beforeOnError(eventsCache::errorSeen)
+                    // terminateOnNextException false -> LB is after this operator, if LB throws do best effort retry
+                    .retryWhen(false, retryStrategy);
+        });
+    }
+
+    private static final class ServiceDiscovererEventsCache<R, E extends ServiceDiscovererEvent<R>> {
+        @SuppressWarnings("rawtypes")
+        private static final Map NONE_RETAINED = emptyMap();
+
+        private final String targetResource;
+        private final UnaryOperator<E> makeUnavailable;
+        private final Map<R, E> currentState = new HashMap<>();
+        private Map<R, E> retainedState = noneRetained();
+
+        private ServiceDiscovererEventsCache(final String targetResource, final UnaryOperator<E> makeUnavailable) {
+            this.targetResource = targetResource;
+            this.makeUnavailable = makeUnavailable;
+        }
+
+        void errorSeen(final Throwable t) {
+            if (retainedState == NONE_RETAINED) {
+                retainedState = new HashMap<>(currentState);
+                currentState.clear();
+            }
+            LOGGER.debug("{} observed an error from ServiceDiscoverer", targetResource, t);
+        }
+
+        Collection<E> consumeAndFilter(final Collection<E> events) {
+            if (retainedState == NONE_RETAINED) {
+                for (E event : events) {
+                    if (UNAVAILABLE.equals(event.status())) {
+                        currentState.remove(event.address());
+                    } else {
+                        currentState.put(event.address(), event);
+                    }
+                }
+                return events;
+            }
+
+            // We have seen an error and re-subscribed upon retry. Based on the Publisher rule 1.10
+            // (https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.10), each subscribe
+            // expects a different Subscriber. Therefore, discovery Publisher suppose to start from a fresh state. We
+            // should populate currentState with new addresses and deactivate the ones which are not present in the new
+            // collection, but were left in retainedState.
+            assert currentState.isEmpty();
+            final List<E> toReturn = new ArrayList<>(events.size() + retainedState.size());
+            for (E event : events) {
+                final R address = event.address();
+                toReturn.add(event);
+                retainedState.remove(address);
+                if (!UNAVAILABLE.equals(event.status())) {
+                    currentState.put(address, event);
+                }
+            }
```

Comment:
I think we want to avoid emitting events that are already known state. In theory we know the status isn't unavailable so we just need to compare the status and metadata but we don't actually know about metadata. Since the events are coming from the same source simple equality should work to capture metadata on a refined `ServiceDiscovererEvent`.
```suggestion
            for (E event : events) {
                final R address = event.address();
                if (UNAVAILABLE.equals(event.status())) {
                    LOGGER.warn("When recovering from ServiceDiscoverer error an UNAVAILABLE event was observed " +
                            "which may indicate that the first event was not a state-of-the-world event and " +
                            "therefore is a ServiceDiscoverer bug.");
                    continue;
                }
                currentState.put(address, event);
                E oldEvent = retainedState.remove(address);
                if (!oldEvent.equals(event)) {
                    toReturn.add(event);
                }
            }
```

Reply:
The `RetryingServiceDiscoverer` should not alter the original flow of events from `ServiceDiscoverer`. We only need to purge addresses after retry to enforce new "state of the world".
I've added tests that validate duplicate events are propagated as-is and clarified that in the comment above this section.

Reply:
I think based on the contract of `ServiceDiscovererEvent` you're technically right, it is legal to do it, but it will seem _odd_ to the consumer (the LB) to receive a no-op message. That said it would mean we'd need to filter out empty results since perhaps the most common result would be the no-op result where nothing changed.

Reply:
It won't receive a noop, it will always receive a new SOTW. See `RetryingServiceDiscovererTest.sameAddressPostRetry()`: https://github.com/apple/servicetalk/pull/3006/files#diff-43e3e0456e8b6d6ddfeeedc106c79ec7ab5a5c2cb13eb325ee91e2cd257ee4d8R104-R122

We don't know if the event changed any meta-data or not. So, after re-subscribe we have to propagate the state as-is. We only prune what is not present in the new state.